File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> result;
        int n = intervals.size();
        
        if(n == 1){
            return intervals;
        }
        
        sort(intervals.begin(), intervals.end());
        
        int start = 0, next = 1;
        vector<int> startv, nextv;
        startv = intervals[start];
        
        for(int i = 1; i < n; i++){
            next = i;
            nextv = intervals[next];
            
            if(startv[1] >= nextv[0]){
                // overlapping
                startv[1] = max(startv[1], nextv[1]);
            }else{
                result.push_back(startv);
                start = i;
                startv = intervals[start];
            }
        }
        
        result.push_back(startv);
        return result;
    }
};

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # sort on the start position
        intervals.sort() 
        ans = []
        n = len(intervals)

        start, end = intervals[0][0], intervals[0][1]
        
        for i in range(1, n):
            # compare with the next start position with the current one
            if intervals[i][0] <= end: 
                end = max(end, intervals[i][1])
            else:
                # then seperate the ones
                ans.append([start, end]) 
                start, end = intervals[i][0], intervals[i][1]

        # add the last remaining ones
        ans.append([start, end])

        return ans

# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna